### PR TITLE
More descriptive error message for diagnostic 727

### PR DESCRIPTION
**Pull request by baronfel**
Will fill in later, my local build isn't working so I'm going to piggyback onto MS's CI :D

The gist is that this error is garbage:
image

We know the expected field names and the positions provided, so we can estimate names not provided and emit them here.  The message can be changed to let the user know this is a suggestion, because DU cases can be added at any point. I'd argue that it's most likely to add them at the end, though.

The new error message is something like

```
This union case expects 3 arguments in tupled form, but was given 2. The missing field arguments may be any of:
  <field1>
  <field2>
  ...
  <fieldN>
```

because that at least gives the user a clue what kinds of fields/data they can expect (without having to rely on tooltips/go-to-definition for the DU)

Open questions:
* [ ] layout of the message itself - right now indentation isn't correct
* [ ] any more intelligence that should be done to try to match patterns provided by the user with the _intended_ DU fields, instead of a basic positional approach that I've looked at for 2 seconds

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
This is amazing, and is related to #1103 if we are going to adjust the error message as well.

**Comment by T-Gro** [comment]
@baronfel , can I help with finalizing this?

To be honest, I do **not** think this should render **"Item2"** in the message if the argument is not named - would rather fallback to the type instead.

Having said that, I would be happy to pick this up from you if you agree.

**Comment by baronfel** [comment]
Absolutely! To be honest I forgot that I submitted this PR, and I'd love someone to pick up the torch.

File: src/Compiler/Checking/CheckPatterns.fs [hunk 1]
```diff
@@ -678,7 +678,13 @@ and TcPatLongIdentUnionCaseOrExnCase warnOnUpper cenv env ad vFlags patEnv ty (m
         elif numArgs < numArgTys then
             if numArgTys > 1 then
                 // Expects tuple without enough args
-                errorR (Error (FSComp.SR.tcUnionCaseExpectsTupledArguments numArgTys, m))
+                let missingArgs = 
+                    argNames.[numArgs..numArgTys - 1]
+                    |> List.map (fun id -> (if id.rfield_name_generated then "" else id.DisplayName + ": ") + id.FormalType.ToString())
+                    |> String.concat "\n, "
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckPatterns.fs:684]: Should line endings be platform specific? I.e. `Environment.NewLine`?

Comment (T-Gro) [reply src/Compiler/Checking/CheckPatterns.fs:684]: In Unix platforms, it is \n anyway.
And in VS on Windows, it works in the rendered popup:

image


But I will modify, sure.

Comment (vzarytovskii) [reply src/Compiler/Checking/CheckPatterns.fs:684]: Also, it's an unusual comma placement. Are we sure we want it there? :)

Comment (T-Gro) [reply src/Compiler/Checking/CheckPatterns.fs:684]: Thanks @nojaf , good catch. Indeed, using NicePrint.minimalStringOfType env.DisplayEnv will be more systematic here.

Comment (T-Gro) [reply src/Compiler/Checking/CheckPatterns.fs:684]: image

**Comment by T-Gro** [comment]
Current look & feel:

image

**Comment by vzarytovskii** [comment]
> Current look & feel:
> 
> image
> 

I feel like this comma placement is awkward. Do we place it in the beginning of the line somewhere else?

**Comment by T-Gro** [comment]
> 

Checked, there are places which use Newline + tab. Newline + comma is not really used, so I am dropping it in favour of tab.

See here:
image

File: tests/service/PatternMatchCompilationTests.fs [hunk 2]
```diff
@@ -249,6 +249,22 @@ match None with
...
+type U =
+    | B of  f1:int seq * {|X:string|} * f3:U * f4: (int * System.String)
+
+let x : U = failwith ""
+match x with
+| B  -> 42
+"""
+    dumpDiagnostics checkResults |> shouldEqual [
+        "(7,2--7,3): This union case expects 4 arguments in tupled form, but was given 0. The missing field arguments may be any of: f1: seq<int> {| X: string |} f3: U f4: (int * System.String)"
+    ]
```
Comment (vzarytovskii) [inline tests/service/PatternMatchCompilationTests.fs:267]: We should be having new tests in the `ComponentTests` suite instead, if they're generic enough (i.e. not service-, autocomplete- or tooling-related).

File: src/Compiler/Checking/CheckPatterns.fs [hunk 3]
```diff
@@ -678,7 +678,14 @@ and TcPatLongIdentUnionCaseOrExnCase warnOnUpper cenv env ad vFlags patEnv ty (m
         elif numArgs < numArgTys then
             if numArgTys > 1 then
                 // Expects tuple without enough args
-                errorR (Error (FSComp.SR.tcUnionCaseExpectsTupledArguments numArgTys, m))
+                let printT = NicePrint.minimalStringOfType env.DisplayEnv
```
[the remark this answers is not in the document]
Comment (T-Gro) [reply src/Compiler/Checking/CheckPatterns.fs:681]: Ok, will do

File: tests/service/PatternMatchCompilationTests.fs (hunk 2 above)
Comment (T-Gro) [reply tests/service/PatternMatchCompilationTests.fs:267]: Ok, will see how this can be migrated

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**